lib/node_modules/@stdlib/repl/base/ctor/lib/main.js

```diff
@@ -0,0 +1,591 @@
+	this.on( 'input', onInput );
+	this.on( 'interrupt', onInterrupt );
+
+	// TODO: check whether to synchronously initialize a REPL history file
+
+	// TODO: check whether to synchronously initialize a REPL log file
+
+	this._count += 1;
+	return this;
+
+	/**
+	* Callback invoked upon encountering a command input.
+	*
+	* @private
+	* @param {string} input - input command
+	*/
+	function onInput( input ) {
+		debug( 'Received an input: %s', input );
+		if ( self._closed === false ) {
+			processInput( self, input );
+		}
+	}
+
+	/**
+	* Callback invoked upon receiving an interrupt.
+	*
+	* @private
+	* @returns {void}
+	*/
+	function onInterrupt() {
+		debug( 'Received an interrupt.' );
+
+		// If the REPL is currently busy executing a command, stop waiting for it to finish:
+		if ( self._busy ) {
+			self._done( new Error( 'unexpected error. Command execution terminated.' ) );
+			return;
+		}
+
+		// Reset the command queue:
+		self._queue.clear();
+	}
+
+	/**
+	* Callback invoked upon executing a command.
+	*
+	* @private
+	* @param {string} cmd - command
+	* @param {boolean} success - boolean indicating whether the command successfully executed
+	* @param {*} res - result
+	* @param {boolean} silent - boolean indicating whether the command output should be silenced
+	*/
+	function onCommand( cmd, success ) {
+		self._count += 1;
+		debug( 'Command count: %d', self._count );
+
+		// Update the internal command history buffer: [..., <id>, <cmd>, <success>, ...]
+		self._history.push( self._count, cmd, success );
+
+		// TODO: if successful and if necessary, (asynchronously?) write the command to a history file (question: do we only want to write successful commands to the history file? maybe we need to option for limiting to successful commands?)
+
+		// TODO: if necessary, (asynchronously?) write the command and result to a log file (JSON serialization?)
+	}
+}
+
+/*
+* Inherit from the `EventEmitter` prototype.
+*/
+inherit( REPL, EventEmitter );
+
+/**
+* Checks if the given input is a possible multi-line command.
+*
+* @private
+* @name _isMultilineInput
+* @memberof REPL.prototype
+* @type {Function}
+* @param {string} input - input command
+* @returns {boolean} boolean indicating whether the given input is a possible multi-line command
+*/
+setNonEnumerableReadOnly( REPL.prototype, '_isMultilineInput', isMultilineInput );
+
+/**
+* Processes a "raw" input command..
+*
+* @private
+* @name _processCommand
+* @memberof REPL.prototype
+* @type {Function}
+* @param {string} code - unevaluated command
+* @returns {(string|Error)} processed command or an error
+*/
+setNonEnumerableReadOnly( REPL.prototype, '_processCommand', processCommand );
+
+/**
+* Compiles a command.
+*
+* @private
+* @name _compileCommand
+* @memberof REPL.prototype
+* @type {Function}
+* @param {string} cmd - command string
+* @returns {(Object|Error)} compiled command or an error
+*/
+setNonEnumerableReadOnly( REPL.prototype, '_compileCommand', compileCommand );
```

Comment:
This is a new private method. Who is the intended caller?

Reply:
I see this is being called by inheritors. If it is meant to be a public method, make it public. Don't force descendants to call private commands. A dev working on this should expect that anything prefixed with a `_` is fair game to change. By having downstream consumers call private methods, you break that contract.

That applies more generally. Simply because a descendant inherits from a parent, that does not mean a descendant should be privy to all the internals of a parent. If a parent classifies something as private, including internal data structures, those should be considered private even to descendants. If you need to provide info/data/methods to descendants, you need to create public APIs/contracts. Otherwise, you are creating tight coupling which is inherently brittle.